### PR TITLE
Optimization suggestions

### PR DESCRIPTION
**Pull request by pedrobergamini**
Hi! Please review and let me know if the following changes make sense (it can be in telegram).

Change 1 - Changed use of memory to storage at _pendingVaultRewards and _pendingYieldRewards, so you save gas by not having to load the deposits array to memory unnecessarily, which might get quite big for users staking big values.

Change 2 - Added _withUpdate bool value to _processVaultRewards, so user.subVaultRewards doesn't get updated twice (wasting gas) when the parent _processRewards function uses _withUpdate false, which means that the top level function (e.g stake, unstake, ...) is already updating the checkpoint values (subYieldRewards and subVaultRewards).

Change 3 (Please review this one carefully) - Fixed updateStakeLock function, the previous fix wasn't updating user.subYieldRewards and user.subVaultRewards after lock period (hence, the user.totalWeight) is changed by the internal _updateStakeLock call. That could lead to invalid rewards calculations for users increasing their lock period.

Change 4 - Switched stake call to synr pool in stakeAsPool to use _useSSYNR as false instead of true. This way, users compounding synr from the lp pool also claim their pending rewards as synr instead of ssynr in the main pool. (Does it make sense? I think that's the intended behaviour)

### CODE REVIEW

**Comment by superpower007** [comment]
Thank you so much for the PR!  I have some questions needs clarify:  

Change 2,  would it be more consistent to always update  subVaultRewards in  _processVaultRewards and update user.subYieldRewards in  _processRewards, and the higher level function always not to do the update? 

Change 4: i am not seeing it in the pr.   yes it is should be false and it should be false from the original code and i don't think we made change to it.

**Comment by sullof** [comment]
(We should add some tests and extend the changes to the flash pool but) the PR looks good to me

**Comment by pedrobergamini** [comment]
> Thank you so much for the PR! I have some questions needs clarify:
> 
> Change 2, would it be more consistent to always update subVaultRewards in _processVaultRewards and update user.subYieldRewards in _processRewards, and the higher level function always not to do the update?
> 
> Change 4: i am not seeing it in the pr. yes it is should be false and it should be false from the original code and i don't think we made change to it.

Hi! Change 4 is present at line 206 of SyndicateCorePool.sol. Regarding Change 2, that can't be done because some times the higher level function might change `user.totalWeight` after `_processRewards()` is called, hence requiring the checkpoint values (subYieldRewards and subVaultRewards) to be updated.

File: contracts/pools/SyndicateCorePool.sol [hunk 1]
```diff
@@ -203,7 +203,7 @@ contract SyndicateCorePool is SyndicatePoolBase {
     _sync();
     User storage user = users[_staker];
     if (user.tokenAmount > 0) {
-      _processRewards(_staker, true, false);
+      _processRewards(_staker, false, false);
```
Comment (pedrobergamini) [inline contracts/pools/SyndicateCorePool.sol:206]: @superpower007

**Comment by pedrobergamini** [comment]
Added one more commit that increases ```REWARD_PER_WEIGHT_MULTIPLIER``` to 1e20 (from 1e12),  in order to avoid rounding rewards to 0 in some ```_sync``` calls. Please just make sure it makes sense with your token supply.

**Review by superpower007 — APPROVED** [review]

**Pull request merged**